SourceCode/raytracer/raytracer/CRTTypes.h

```diff
@@ -0,0 +1,335 @@
+        Vec3 result;
+        result.x = (*this)(0, 0) * vec.x + (*this)(0, 1) * vec.y + (*this)(0, 2) * vec.z;
+        result.y = (*this)(1, 0) * vec.x + (*this)(1, 1) * vec.y + (*this)(1, 2) * vec.z;
+        result.z = (*this)(2, 0) * vec.x + (*this)(2, 1) * vec.y + (*this)(2, 2) * vec.z;
+        return result;
+    }
+
+    Matrix3x3& operator*=(const Matrix3x3& other) {
+        *this = *this * other;
+        return *this;
+    }
+
+    Vec3 col(uint8_t num) const
+    {
+        if (num > 2) {
+            throw std::out_of_range("Matrix3x3::col: Index out of range: " + num);
+        }
+        return Vec3{ data[num], data[num + 3], data[num + 6] };
+    }
+
+    static Matrix3x3 identity() {
+        return Matrix3x3();
+    }
+
+    static Matrix3x3 translation(float tx, float ty) {
+        Matrix3x3 result = identity();
+        result(0, 2) = tx;
+        result(1, 2) = ty;
+        return result;
+    }
+
+    static Matrix3x3 scaling(float sx, float sy) {
+        Matrix3x3 result = identity();
+        result(0, 0) = sx;
+        result(1, 1) = sy;
+        return result;
+    }
+
+    static Matrix3x3 rotation2D(float angle) {
+        Matrix3x3 result = identity();
+        float c = std::cos(angle);
+        float s = std::sin(angle);
+        result(0, 0) = c;
+        result(0, 1) = -s;
+        result(1, 0) = s;
+        result(1, 1) = c;
+        return result;
+    }
+
+    /* Rotate around X. pitch */
+    static Matrix3x3 Tilt(float deg) {
+        float rad = radFromDeg * deg;
+        float cosv = cos(rad);
+        float sinv = sin(rad);
+
+        return Matrix3x3 {{
+        1.f, 0.f, 0.f,
+        0.f, cosv, -sinv,
+        0.f, sinv, cosv
+        }};
+    }
+
+    /* Rotate around Y. yaw */
+    static Matrix3x3 Pan(float deg) {
+        float rad = radFromDeg * deg;
+        float cosv = cos(rad);
+        float sinv = sin(rad);
+
+        return Matrix3x3 {{
+        cosv,  0, sinv,
+        0,     1, 0,
+        -sinv, 0, cosv
+        }};
+    }
+
+    /*
+    * Rotate around Z
+    */
+    static Matrix3x3 Roll(float deg) {
+        float rad = radFromDeg * deg;
+        float cosv = cos(rad);
+        float sinv = sin(rad);
+
+        return Matrix3x3 {{
+        cosv, -sinv, 0,
+        sinv, cosv, 0,
+        0,0, 1
+        }};
+    }
+
+
+
+    void toString() const {
+        std::string result = "";
+        for (int row = 0; row < 3; ++row) {
+            for (int col = 0; col < 3; ++col) {
+                result += std::to_string((*this)(row, col));
+                result += " ";
+            }
+            result += "\n";
+        }
+    }
+
+private:
+    float data[9];
+    static constexpr float radFromDeg = static_cast<float>(std::numbers::pi) / 180.0f;
+};
+
+struct Ray {
+    Vec3 origin;
+    Vec3 direction;
+
+    Ray() : origin(Vec3()), direction(Vec3()) {}
+
+    Ray(const Vec3& origin, const Vec3& direction) : origin(origin), direction(direction) {}
+};
```

Comment:
Instead of having several different ray types you could use a type field in the Ray itself